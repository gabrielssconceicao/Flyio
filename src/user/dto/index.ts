import { CreateUserDto } from './create-user.dto';
import { FindAllUsersResponseDto } from './find-all-users.dto';
import { QueryParamDto } from './query-param.dto';
import { ReactivateUserDto } from './reactivate-user.dto';
import { UpdateUserDto } from './update-user.dto';

export {
  CreateUserDto,
  FindAllUsersResponseDto,
  QueryParamDto,
  ReactivateUserDto,
  UpdateUserDto,
};
