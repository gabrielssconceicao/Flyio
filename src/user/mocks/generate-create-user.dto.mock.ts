import { CreateUserDto } from '../dto';
import { userMock } from './user.mock';
export const generateCreateUserDtoMock = (
  bio: boolean = false,
): CreateUserDto => {
  const dto: CreateUserDto = {
    name: userMock.name,
    username: userMock.username,
    email: userMock.email,
    password: userMock.password,
  };

  if (bio) dto.bio = userMock.bio;
  return dto;
};
