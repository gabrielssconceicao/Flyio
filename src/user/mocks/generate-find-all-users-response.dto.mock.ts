import { FindAllUsersResponseDto } from '../dto';
import { userMock } from './user.mock';

export const generateFindAllUsersResponseDtoMock =
  (): FindAllUsersResponseDto => {
    return {
      count: 1,
      items: [
        {
          id: userMock.id,
          name: userMock.name,
          username: userMock.username,
          profileImg: userMock.profileImg,
        },
      ],
    };
  };
