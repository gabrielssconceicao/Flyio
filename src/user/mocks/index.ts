import { generateUserMock } from './user.mock';
import { generateFindAllUsersResponseDtoMock } from './generate-find-all-users-response.dto.mock';
import { generateCreateUserDtoMock } from './generate-create-user.dto.mock';

export {
  generateUserMock,
  generateFindAllUsersResponseDtoMock,
  generateCreateUserDtoMock,
};
