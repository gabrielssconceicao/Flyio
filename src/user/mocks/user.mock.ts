import { User } from '../entities/user.entity';

export const userMock = {
  id: '42-d-f-df4',
  name: 'John Doe',
  username: 'jDoe453',
  email: 'jdoe@me.com',
  password: '123456',
  profileImg: 'https://example.com/profile.jpg',
  bio: 'This is my bio',
  active: true,
};
export const generateUserMock = (): User => {
  return {
    id: userMock.id,
    name: userMock.name,
    username: userMock.username,
    profileImg: userMock.profileImg,
    bio: userMock.bio,
    active: userMock.active,
    followers: 0,
    following: 0,
  };
};
