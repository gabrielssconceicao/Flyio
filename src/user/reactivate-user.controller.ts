import { Controller, HttpCode, HttpStatus, Param, Patch } from '@nestjs/common';
import { ApiBody, ApiOperation, ApiResponse, ApiTags } from '@nestjs/swagger';
import { ReactivateUserService } from './reactivate-user.service';
import { ReactivateUserDto } from './dto';

@ApiTags('Users')
@Controller('users')
export class ReactivateUserController {
  constructor(private readonly reactivateUserService: ReactivateUserService) {}

  @ApiOperation({ summary: 'Reactivate user by token' })
  @ApiBody({
    type: ReactivateUserDto,
    description: 'Reactivate user by token',
  })
  @ApiResponse({
    status: HttpStatus.CREATED,
    description: 'Account reactivated successfully.',
    schema: {
      example: {
        message: 'Account reactivated successfully',
      },
    },
  })
  @ApiResponse({
    status: HttpStatus.UNAUTHORIZED,
    description: 'Invalid or expired token.',
    schema: {
      example: {
        message: 'Invalid or expired token',
      },
    },
  })
  @HttpCode(HttpStatus.CREATED)
  @Patch('reactivate/:token')
  reactivate(@Param() reactivateUserDto: ReactivateUserDto) {
    return this.reactivateUserService.reactivate(reactivateUserDto);
  }
}
