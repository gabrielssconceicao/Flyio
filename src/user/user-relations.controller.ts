import {
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  Patch,
  Query,
  UseGuards,
} from '@nestjs/common';
import {
  ApiBody,
  ApiOperation,
  ApiParam,
  ApiResponse,
  ApiTags,
} from '@nestjs/swagger';
import { ReactivateUserDto } from './dto';
import { UserRelationsService } from './user-relations.service';
import { PaginationDto } from '../common/dto/pagination.dto';
import { AuthTokenGuard } from '../auth/guard/auth-token.guard';
import { ApiAuthResponses } from '../common/decorators/guard.decorator';
import { FindAllUsersPostsResponseDto } from './dto/find-all-user-posts.dto';

@ApiTags('Users')
@Controller('users')
export class UserRelationsController {
  constructor(private readonly userRelationsService: UserRelationsService) {}

  @ApiOperation({ summary: 'Reactivate user by token' })
  @ApiBody({
    type: ReactivateUserDto,
    description: 'Reactivate user by token',
  })
  @ApiResponse({
    status: HttpStatus.CREATED,
    description: 'Account reactivated successfully.',
    schema: {
      example: {
        message: 'Account reactivated successfully',
      },
    },
  })
  @ApiResponse({
    status: HttpStatus.UNAUTHORIZED,
    description: 'Invalid or expired token.',
    schema: {
      example: {
        message: 'Invalid or expired token',
      },
    },
  })
  @HttpCode(HttpStatus.CREATED)
  @Patch('reactivate/:token')
  reactivate(@Param() reactivateUserDto: ReactivateUserDto) {
    return this.userRelationsService.reactivate(reactivateUserDto);
  }

  @ApiOperation({ summary: 'Get all posts of a user by username' })
  @ApiParam({
    name: 'username',
    description: 'Username of the user',
    example: 'jDoe453',
  })
  @ApiResponse({
    status: HttpStatus.OK,
    description: 'List of posts of a user with limited details',
    type: FindAllUsersPostsResponseDto,
  })
  @ApiAuthResponses()
  @HttpCode(HttpStatus.OK)
  @UseGuards(AuthTokenGuard)
  @Get(':username/posts')
  getAllPostsByUsername(
    @Param('username') username: string,
    @Query() paginationDto: PaginationDto,
  ) {
    return this.userRelationsService.getAllPostsByUsername(
      username,
      paginationDto,
    );
  }
}
