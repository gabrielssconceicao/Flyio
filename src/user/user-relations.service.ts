import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { FindAllUsersResponseDto } from './dto';
import { PaginationDto } from '../common/dto/pagination.dto';
import { FindAllUsersPostsResponseDto } from './dto/find-all-user-posts.dto';
import { FindAllLikedPostsResponseDto } from '../post/dto/find-all-liked-post.dto';
import { TokenPayloadDto } from '../auth/dto';

@Injectable()
export class UserRelationsService {
  constructor(private readonly prismaService: PrismaService) {}

  private selectPostFields = {
    id: true,
    content: true,
    createdAt: true,
    images: {
      select: {
        url: true,
        id: true,
      },
    },
    user: {
      select: {
        username: true,
        profileImg: true,
      },
    },
    _count: {
      select: {
        PostLikes: true,
      },
    },
  };

  async getAllPostsByUsername(
    username: string,
    paginationDto: PaginationDto,
    tokenPayload: TokenPayloadDto,
  ): Promise<FindAllUsersPostsResponseDto> {
    const { limit = 50, offset = 0 } = paginationDto;

    await this.userExists(username);

    const { count, items } = await this.prismaService.findAll(
      this.prismaService.post,
      {
        where: { user: { username } },
        select: {
          ...this.selectPostFields,
          PostLikes: {
            where: { userId: tokenPayload.sub },
            select: { userId: true },
          },
        },
        orderBy: { createdAt: 'desc' },
        skip: offset,
        take: limit,
      },
    );

    const mappedPosts = items.map(({ _count, PostLikes, ...post }) => ({
      ...post,
      likes: _count.PostLikes,
      liked: !!PostLikes.length,
    }));

    return { count, items: mappedPosts };
  }

  async getAllLikedPostsByUsername(
    username: string,
    paginationDto: PaginationDto,
  ): Promise<FindAllLikedPostsResponseDto> {
    const { limit = 50, offset = 0 } = paginationDto;
    const user = await this.userExists(username);

    const { count, items } = await this.prismaService.findAll(
      this.prismaService.post,
      {
        where: {
          PostLikes: {
            some: {
              userId: user.id,
            },
          },
        },
        orderBy: {
          createdAt: 'desc',
        },
        take: limit,
        skip: offset,
        select: this.selectPostFields,
      },
    );

    return {
      count,
      items: items.map(({ _count, ...post }) => {
        return { ...post, likes: _count.PostLikes, liked: true };
      }),
    };
  }

  async followUser(username: string, tokenPayload: TokenPayloadDto) {
    const followingUser = await this.userExists(username);

    if (!followingUser) {
      throw new NotFoundException('User not found');
    }

    if (followingUser.id === tokenPayload.sub) {
      throw new BadRequestException('You cannot follow yourself');
    }

    const follower = await this.prismaService.follower.findFirst({
      where: {
        followingId: followingUser.id,
        userId: tokenPayload.sub,
      },
    });

    if (follower) {
      throw new BadRequestException('User already followed');
    }

    await this.prismaService.follower.create({
      data: {
        followingId: followingUser.id,
        userId: tokenPayload.sub,
      },
    });
  }

  async unfollowUser(username: string, tokenPayload: TokenPayloadDto) {
    const followingUser = await this.userExists(username);

    if (followingUser.id === tokenPayload.sub) {
      throw new BadRequestException('You cannot unfollow yourself');
    }

    const follower = await this.prismaService.follower.findFirst({
      where: {
        followingId: followingUser.id,
        userId: tokenPayload.sub,
      },
    });

    if (!follower) {
      throw new BadRequestException('User already unfollowed');
    }

    await this.prismaService.follower.delete({
      where: {
        userId_followingId: {
          followingId: followingUser.id,
          userId: tokenPayload.sub,
        },
      },
    });
  }
  async getAllFollowingsByUser(
    username: string,
  ): Promise<FindAllUsersResponseDto> {
    const user = await this.userExists(username);
    if (!user) {
      throw new BadRequestException('User not found');
    }

    const { count, items } = await this.prismaService.findAll(
      this.prismaService.follower,
      {
        where: { userId: user.id },
        select: {
          following: {
            select: {
              username: true,
              name: true,
              profileImg: true,
              id: true,
            },
          },
        },
      },
    );

    return { count, items: items.map((item) => item.following) };
  }
  async getAllFollowersByUser(
    username: string,
  ): Promise<FindAllUsersResponseDto> {
    const user = await this.userExists(username);
    if (!user) {
      throw new BadRequestException('User not found');
    }

    const { count, items } = await this.prismaService.findAll(
      this.prismaService.follower,
      {
        where: { followingId: user.id },
        select: {
          user: {
            select: {
              username: true,
              name: true,
              profileImg: true,
              id: true,
            },
          },
        },
      },
    );

    return { count, items: items.map((item) => item.user) };
  }

  private async userExists(username: string) {
    const user = await this.prismaService.user.findUnique({
      where: { username },
    });

    if (!user) {
      throw new NotFoundException('User not found');
    }

    return user;
  }
}
