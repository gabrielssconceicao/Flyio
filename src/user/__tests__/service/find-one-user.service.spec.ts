import { Test, TestingModule } from '@nestjs/testing';
import { JwtService } from '@nestjs/jwt';
import { NotFoundException } from '@nestjs/common';
import { User } from '../../entities/user.entity';
import { generateUserMock } from '../../mocks';
import { UserService } from '../../user.service';
import { HashingServiceProtocol } from 'src/auth/hashing/hashing.service';
import { hashingServiceMock, jwtServiceMock } from 'src/auth/mocks';
import { PrismaService } from 'src/prisma/prisma.service';
import { userPrismaService } from 'src/prisma/mock/prisma.service.mock';
import { CloudinaryService } from 'src/cloudinary/cloudinary.service';
import { cloudinaryServiceMock } from 'src/cloudinary/mocks';
import { PermissionService } from 'src/permission/permission.service';

describe('<UserService - (findOne) />', () => {
  let service: UserService;
  let hashingService: HashingServiceProtocol;
  let prismaService: PrismaService;
  let cloudinaryService: CloudinaryService;

  let user: User;
  let username: string;
  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        UserService,
        {
          provide: HashingServiceProtocol,
          useValue: hashingServiceMock,
        },
        {
          provide: PrismaService,
          useValue: userPrismaService,
        },
        {
          provide: CloudinaryService,
          useValue: cloudinaryServiceMock,
        },
        {
          provide: JwtService,
          useValue: jwtServiceMock,
        },
        {
          provide: PermissionService,
          useValue: {},
        },
      ],
    }).compile();

    service = module.get<UserService>(UserService);
    hashingService = module.get<HashingServiceProtocol>(HashingServiceProtocol);
    prismaService = module.get<PrismaService>(PrismaService);
    cloudinaryService = module.get<CloudinaryService>(CloudinaryService);

    user = generateUserMock();
    username = user.username;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
    expect(hashingService).toBeDefined();
    expect(prismaService).toBeDefined();
    expect(cloudinaryService).toBeDefined();
  });

  it('should return a user successfully', async () => {
    jest.spyOn(prismaService.user, 'findUnique').mockResolvedValue({
      ...user,
      _count: { followers: 0, following: 0 },
    } as any);

    const result = await service.findOne(username);

    expect(prismaService.user.findUnique).toHaveBeenCalled();

    expect(result).toEqual(result);
    expect(result).toMatchSnapshot();
  });

  it('should throw an NotFoundException', async () => {
    jest.spyOn(prismaService.user, 'findUnique').mockResolvedValue(null);

    await expect(service.findOne(username)).rejects.toThrow(NotFoundException);
  });
});
