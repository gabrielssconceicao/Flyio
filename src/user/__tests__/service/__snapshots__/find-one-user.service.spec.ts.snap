// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserService - (findOne) /> should return a user successfully 1`] = `
{
  "active": true,
  "bio": "This is my bio",
  "followers": 0,
  "following": 0,
  "id": "42-d-f-df4",
  "name": "John Doe",
  "profileImg": "https://example.com/profile.jpg",
  "username": "jDoe453",
}
`;
